### PR TITLE
[sonic-swss] Add support for COPP

### PR DESCRIPTION
**Pull request by lguohan**
* Added new class CoppOrch
* Added schema for COPP_TABLE

Signed-off-by: hrachya@mellanox.com

### CODE REVIEW

File: orchagent/copporch.cpp [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+        }
+
+        if (m_trap_map.find(key) != m_trap_map.end())
+        {
+            SWSS_LOG_DEBUG("found existing trap_group object:%s", key.c_str());
+            if(!policer_attribs.empty())
+            {
+                attr.id = SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER;
+                attr.value.oid = SAI_NULL_OBJECT_ID;
+                if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->get_trap_group_attribute(m_trap_map[key], 1, &attr)))
```
Comment (lguohan) [inline orchagent/copporch.cpp]: It is not necessary to use get api to retrieve the trap group ID from the SAI as you know if you have created the trap group or not, and should cache the trap group ID. In the orchagent, we should not use get API unless we have to. For example, the hw_lane ID, number of ports, and etc. on. Everything else should be cached in the orchagent.

File: orchagent/copporch.cpp [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+                {
+                    SWSS_LOG_ERROR("Failed to get policer for trap gorup:%s, error:%d\n", key.c_str(), sai_status);
```
Comment (lguohan) [inline orchagent/copporch.cpp]: typo -> trap group

File: orchagent/copporch.cpp [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+                    return task_process_status::task_failed;
+                }
+                SWSS_LOG_DEBUG("Obtained policer object:%llx for trap_group:%s", attr.value.oid, key.c_str());
+                policer_id = attr.value.oid;                
+                SWSS_LOG_DEBUG("Applying settings to existing policer:%llx for trap_group:%llx, name:%s", policer_id, m_trap_map[key], key.c_str());
+                if(SAI_NULL_OBJECT_ID == policer_id)
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >if( [](start = 16, length = 3)

leave a space here, need to be consistent

File: orchagent/copporch.cpp [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+                {
+                    SWSS_LOG_WARN("Creating policer for existing Trap group:%llx (name:%s).", m_trap_map[key], key.c_str());
+                    if (SAI_STATUS_SUCCESS != (sai_status = sai_policer_api->create_policer(&policer_id, policer_attribs.size(), policer_attribs.data())))
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >policer_id [](start = 93, length = 10)

once you have created the policer_id, I do not see how you assign the policer_id to the trap_group.

File: orchagent/copporch.cpp [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+}
+
+bool CoppOrch::removePolicerFromTrapGroup(sai_object_id_t trap_group)
+{
+    sai_attribute_t attr;
+    sai_status_t sai_status;
+    sai_object_id_t policer_id;    
+    attr.id = SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER;
+    attr.value.oid = SAI_NULL_OBJECT_ID;
+    if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->get_trap_group_attribute(trap_group, 1, &attr))) {
```
Comment (lguohan) [inline orchagent/copporch.cpp]: > if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->get_trap_group_attribute(trap_group, 1, &attr))) { [](start = 3, length = 107)

do not use get unless you have to.

File: orchagent/copporch.cpp [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+        SWSS_LOG_ERROR("Failed to get policer for trap gorup:%llx, error:%d\n", trap_group, sai_status);
+        return false;
+    }
+    policer_id = attr.value.oid;
+    if(SAI_NULL_OBJECT_ID == policer_id)
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >if( [](start = 4, length = 3)

leave a space here

File: orchagent/copporch.cpp [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+    attr.value.oid = SAI_NULL_OBJECT_ID;
+    if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->set_trap_group_attribute(trap_group, &attr))) {
+        SWSS_LOG_ERROR("Failed to reset policer for trap gorup:%llx, error:%d\n", trap_group, sai_status);
+        return false;
+    }
+    if (SAI_STATUS_SUCCESS != (sai_status = sai_policer_api->remove_policer(policer_id)))
+    {
+        SWSS_LOG_ERROR("Failed to remove policer:%llx, error:%d\n", policer_id, sai_status);
+        return false;
+    }
```
Comment (lguohan) [inline orchagent/copporch.cpp]: is it possible the policer_id is used by multiple trap groups?

File: orchagent/copporch.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+protected:
+    virtual void doTask(Consumer& consumer);
+    task_process_status processCoppRule(Consumer& consumer);
+    bool isValidList(vector<string> &trap_id_list, vector<string> &all_items) const;
+    bool getTrapID(string &trap_id_str, sai_hostif_trap_id_t &trap_id) const;
+    bool getTrapIdList(vector<string> &trap_id_name_list, vector<sai_hostif_trap_id_t> &trap_id_list) const;
+    bool applyTrapIds(sai_object_id_t trap_group, vector<string> &trap_id_name_list, std::vector<sai_attribute_t> &trap_id_attribs);
+    bool removePolicerFromTrapGroup(sai_object_id_t trap_group);
+protected:
+    object_map m_trap_map;
```
Comment (lguohan) [inline orchagent/copporch.h]: >m_trap_map [](start = 15, length = 10)

rename to m_trap_group_map as it stores the trap_group to id map?

File: orchagent/copporch.cpp [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+                SWSS_LOG_DEBUG("Bound policer to the trap group");
+            }
+        }
+
+        //
+        // The channel is always the NETDEV device.
+        //
+        attr.id = SAI_HOSTIF_TRAP_ATTR_TRAP_CHANNEL;
+        attr.value.s32 = SAI_HOSTIF_TRAP_CHANNEL_NETDEV;
+        trap_id_attribs.push_back(attr);
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >trap_id_attribs [](start = 8, length = 15)

you are building trap_id_attribs in multiple places. 

Can you consolidate the attributes building in one place. It's easier for readability. 

here is my suggestion:

1. get key values from db, fill your own trap object.

e.g, map<trap_id, tuple<trap_group_name, trap_action>>

2. In one function, translate the trap object into sai trap object and call corresponding sai function.

set trap_id 
set trap_channel
set trap_group
call set_trap_attribute()

Comment (lguohan) [inline orchagent/copporch.cpp]: I feel if you first create an internal trap object, and then translate the internal trap object into sai trap object, it maybe easier to read.

---
In reply to: [73782200](https://github.com/Azure/sonic-swss/pull/57#discussion_r73782200) [](ancestors = 73782200)

**Comment by lguohan** [comment]
@hrachya-mellanox , @stcheng

File: orchagent/copporch.cpp (hunk 7 above)
Comment (hrachya-mellanox) [inline orchagent/copporch.cpp]: no it shouldn't be possible.

**Comment by hrachya-mellanox** [comment]
I've addressed Guohan's comments:
- styling
- policer_id is bound to the trap_group, for case when adding policer to existing trap group.
- removed get__attrib calls. Instead policer values are cached now.
- consolidated trap_attrs in one function.

File: orchagent/copporch.cpp [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+    return true;
+}
+
+sai_object_id_t CoppOrch::getTrapGroupPolicer(string key)
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_DEBUG("trap group name:%s:", key.c_str());
+    if (m_trap_group_map.find(key) == m_trap_group_map.end())
+    {
+        return SAI_NULL_OBJECT_ID;
```
Comment (stcheng) [inline orchagent/copporch.cpp]: SAI error codes shall only be returned in the SAI implementation not here.

File: orchagent/copporch.cpp [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+    }
+    SWSS_LOG_DEBUG("trap group id:%llx", m_trap_group_map[key]);
+    if (m_trap_group_policer_map.find(m_trap_group_map[key]) == m_trap_group_policer_map.end())
+    {
+        return SAI_NULL_OBJECT_ID;
```
Comment (stcheng) [inline orchagent/copporch.cpp]: Same as above.

File: orchagent/copporch.cpp [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+    attr.value.s32 = SAI_HOSTIF_TRAP_CHANNEL_NETDEV;
+    trap_id_attribs.push_back(attr);
+
+    for (auto trap_id : trap_id_list)
+    {
+        SWSS_LOG_DEBUG("set trap attr.1:%d", attr.id);
+        for(auto trap_id_attr : trap_id_attribs)
+        {
+            SWSS_LOG_DEBUG("set trap attr.2:%d", trap_id_attr.id);
+            if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->set_trap_attribute(trap_id, &trap_id_attr)))
```
Comment (stcheng) [inline orchagent/copporch.cpp]: Could you align the previous code as to split this into two lines: one for the function call and another for the if. Same as below.

File: orchagent/copporch.cpp [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+    string key = kfvKey(tuple);
+    string op = kfvOp(tuple);
+    
+
+    SWSS_LOG_DEBUG("copp:processing:%s", key.c_str());
+    if (consumer.m_consumer->getTableName() != APP_COPP_TABLE_NAME)
+    {
+        SWSS_LOG_ERROR("Key with invalid table type passed in %s, expected:%s\n", key.c_str(), APP_COPP_TABLE_NAME);
+        return task_process_status::task_invalid_entry;
+    }
```
Comment (stcheng) [inline orchagent/copporch.cpp]: the above check is not needed as it is already checked in the common library

File: orchagent/copporch.cpp [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+            it = consumer.m_toSync.erase(it);
+            break;
+        case task_process_status::task_failed:
+            SWSS_LOG_ERROR("Processing copp task item failed, exiting. ");
+            dumpTuple(consumer, tuple);
+            return;
+        case task_process_status::task_need_retry:
+            SWSS_LOG_ERROR("Processing copp task item failed, will retry.");
+            dumpTuple(consumer, tuple);
+            it++;
```
Comment (stcheng) [inline orchagent/copporch.cpp]: always have a break after a case and have a default at the end

File: orchagent/main.cpp [hunk 15]
```diff
@@ -84,6 +85,8 @@ void initSaiApi()
     sai_api_query(SAI_API_NEXT_HOP_GROUP,       (void **)&sai_next_hop_group_api);
     sai_api_query(SAI_API_ROUTE,                (void **)&sai_route_api);
     sai_api_query(SAI_API_LAG,                  (void **)&sai_lag_api);
+    sai_api_query(SAI_API_POLICER,              (void **)&sai_policer_api);
+
```
Comment (stcheng) [inline orchagent/main.cpp]: extra empty line

File: orchagent/orch.cpp [hunk 16]
```diff
@@ -1,6 +1,7 @@
 #include "orch.h"
 #include "logger.h"
-
+#include <sstream>
```
Comment (stcheng) [inline orchagent/orch.cpp]: i think sstream is not needed

File: orchagent/orch.h [hunk 17]
```diff
@@ -15,6 +15,21 @@ extern "C" {
...
+    task_success,
+    task_invalid_entry,
+    task_failed,
+    task_need_retry
+} task_process_status;
+
+typedef std::map<string, sai_object_id_t> object_map;
+typedef std::pair<string, sai_object_id_t> object_map_pair;
+
+
```
Comment (stcheng) [inline orchagent/orch.h]: one more empty line

File: orchagent/copporch.h [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    CoppOrch(DBConnector *db, string tableName);
+protected:
+    virtual void doTask(Consumer& consumer);
+    task_process_status processCoppRule(Consumer& consumer);
+    bool isValidList(vector<string> &trap_id_list, vector<string> &all_items) const;
+    bool getTrapID(string &trap_id_str, sai_hostif_trap_id_t &trap_id) const;
+    bool getTrapIdList(vector<string> &trap_id_name_list, vector<sai_hostif_trap_id_t> &trap_id_list) const;
+    bool applyTrapIds(sai_object_id_t trap_group, vector<string> &trap_id_name_list, std::vector<sai_attribute_t> &trap_id_attribs);
+    bool removePolicerFromTrapGroup(string trap_group_name);
+protected:
```
Comment (stcheng) [inline orchagent/copporch.h]: why two protected

File: orchagent/copporch.h [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+class CoppOrch : public Orch
+{
+public:
+    CoppOrch(DBConnector *db, string tableName);
+protected:
+    virtual void doTask(Consumer& consumer);
+    task_process_status processCoppRule(Consumer& consumer);
+    bool isValidList(vector<string> &trap_id_list, vector<string> &all_items) const;
+    bool getTrapID(string &trap_id_str, sai_hostif_trap_id_t &trap_id) const;
```
Comment (stcheng) [inline orchagent/copporch.h]: name of the function shall be getTrapId. d is not capitalized.

File: orchagent/copporch.cpp [hunk 20]
```diff
@@ -0,0 +1,488 @@
...
+
+bool CoppOrch::getTrapIdList(vector<string> &trap_id_name_list, vector<sai_hostif_trap_id_t> &trap_id_list) const
+{
+    SWSS_LOG_ENTER();
+    for(auto trap_id_str : trap_id_name_list)
+    {
+        sai_hostif_trap_id_t trap_id;
+        if (!getTrapID(trap_id_str, trap_id))
+        {
+            SWSS_LOG_ERROR("Invalid trap_id value specified %s\n", trap_id_str.c_str());
```
Comment (stcheng) [inline orchagent/copporch.cpp]: this error message will appear twice both in getTrapID function and here. Is it better to combine these two functions and just use try/catch logic to make it simpler?

File: orchagent/copporch.cpp [hunk 21]
```diff
@@ -0,0 +1,488 @@
...
+    trap_id_attribs.push_back(attr);
+
+    // The channel is always the NETDEV device.
+    attr.id = SAI_HOSTIF_TRAP_ATTR_TRAP_CHANNEL;
+    attr.value.s32 = SAI_HOSTIF_TRAP_CHANNEL_NETDEV;
+    trap_id_attribs.push_back(attr);
+
+    for (auto trap_id : trap_id_list)
+    {
+        SWSS_LOG_DEBUG("set trap attr.1:%d", attr.id);
```
Comment (stcheng) [inline orchagent/copporch.cpp]: will this always print the attr.id = SAI_HOSTIF_TRAP_ATTR_TRAP_CHANNEL?

File: orchagent/copporch.cpp [hunk 22]
```diff
@@ -0,0 +1,488 @@
...
+        for(auto trap_id_attr : trap_id_attribs)
+        {
+            SWSS_LOG_DEBUG("set trap attr.2:%d", trap_id_attr.id);
```
Comment (stcheng) [inline orchagent/copporch.cpp]: these two debug msgs could be removed

File: orchagent/copporch.h [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+    bool getTrapIdList(vector<string> &trap_id_name_list, vector<sai_hostif_trap_id_t> &trap_id_list) const;
+    bool applyTrapIds(sai_object_id_t trap_group, vector<string> &trap_id_name_list, std::vector<sai_attribute_t> &trap_id_attribs);
+    bool removePolicerFromTrapGroup(string trap_group_name);
```
Comment (stcheng) [inline orchagent/copporch.h]: removePolicer is a simpler name?

File: orchagent/copporch.h [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+protected:
+    sai_object_id_t getTrapGroupPolicer(string trap_group_name);
+    bool createTrapGroupPolicer(string trap_group, std::vector<sai_attribute_t> &policer_attribs);
```
Comment (stcheng) [inline orchagent/copporch.h]: The above functions can be:
removePolicer
getPolicer
addPolicer

See other orch header files as references.

File: orchagent/copporch.cpp [hunk 25]
```diff
@@ -0,0 +1,488 @@
...
+            task_status = processCoppRule(consumer);
+        }
+        catch(const std::out_of_range e)
+        {
+            SWSS_LOG_ERROR("processing copp rule threw exception:%s", e.what());
+            task_status = task_process_status::task_invalid_entry;            
+        }
+        switch(task_status)
+        {
+        case task_process_status::task_success :
```
Comment (stcheng) [inline orchagent/copporch.cpp]: add four more spaces for the code inside the switch?

File: orchagent/copporch.cpp (hunk 10 above)
Comment (hrachya-mellanox) [inline orchagent/copporch.cpp]: This is 'null' value, not error code. it's used by caller function to detect whether there is cached object.

Comment (stcheng) [inline orchagent/copporch.cpp]: I see. So my suggestion is that similar to the neighorch.h, we will have a hasPolicer() function and getPolicerId() function, which seems to be clearer.

**Comment by lguohan** [comment]
In portsOrch.cpp, line 42 - line 72, should be moved to in this copp orch

File: orchagent/copporch.cpp [hunk 26]
```diff
@@ -0,0 +1,474 @@
...
+CoppOrch::CoppOrch(DBConnector *db, string tableName) :
+    Orch(db, tableName)
+{
+    SWSS_LOG_ENTER();
+};
+
+void CoppOrch::getTrapIdList(vector<string> &trap_id_name_list, vector<sai_hostif_trap_id_t> &trap_id_list) const
+{
+    SWSS_LOG_ENTER();
+    for(auto trap_id_str : trap_id_name_list)
```
Comment (lguohan) [inline orchagent/copporch.cpp]: leave a space between for and '(', make this consistent across the code.

File: orchagent/copporch.cpp [hunk 27]
```diff
@@ -0,0 +1,474 @@
...
+    {"mld_v1_v2", SAI_HOSTIF_TRAP_ID_IPV6_MLD_V1_V2},
+    {"mld_v1_report", SAI_HOSTIF_TRAP_ID_IPV6_MLD_V1_REPORT},
+    {"mld_v2_done", SAI_HOSTIF_TRAP_ID_IPV6_MLD_V1_DONE},
+    {"mld_v2_report", SAI_HOSTIF_TRAP_ID_MLD_V2_REPORT},
+    {"ip2me", SAI_HOSTIF_TRAP_ID_IP2ME},
+    {"ssh", SAI_HOSTIF_TRAP_ID_SSH},
+    {"snmp", SAI_HOSTIF_TRAP_ID_SNMP},
+    {"router_custom_range", SAI_HOSTIF_TRAP_ID_ROUTER_CUSTOM_RANGE_BASE},
+    {"l3_mtu_error", SAI_HOSTIF_TRAP_ID_L3_MTU_ERROR},
+    {"ttl", SAI_HOSTIF_TRAP_ID_TTL_ERROR}
```
Comment (lguohan) [inline orchagent/copporch.cpp]: ttl -> ttl_error

File: orchagent/copporch.cpp [hunk 28]
```diff
@@ -0,0 +1,474 @@
...
+    {
+        SWSS_LOG_ERROR("Failed to create policer for existing trap_group:%llx, name:%s, error:%d", m_trap_group_map[trap_group], trap_group.c_str(), sai_status);
+        return false;
+    }                   
+    SWSS_LOG_DEBUG("Created policer:%llx for trap group:%llx", policer_id, m_trap_group_map[trap_group]);
+    sai_attribute_t attr;
+    attr.id = SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER;
+    attr.value.oid = policer_id;
+    sai_status = sai_hostif_api->set_trap_group_attribute(m_trap_group_map[trap_group], &attr);
+    if (SAI_STATUS_SUCCESS != sai_status) {
```
Comment (lguohan) [inline orchagent/copporch.cpp]: > { [](start = 41, length = 2)

let's put the '{' in separate line, make the style consistent.

File: orchagent/copporch.cpp [hunk 29]
```diff
@@ -0,0 +1,474 @@
...
+}
+task_process_status CoppOrch::processCoppRule(Consumer& consumer)
+{
+    SWSS_LOG_ENTER();
+    sai_status_t sai_status;
+    vector<string> trap_id_list;
+    string queue_ind;
+    auto it = consumer.m_toSync.begin();
+    KeyOpFieldsValuesTuple tuple = it->second;    
+    string key = kfvKey(tuple);
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >key [](start = 11, length = 3)

can we rename this var to trap_group_name for readability?

File: orchagent/copporch.cpp [hunk 30]
```diff
@@ -0,0 +1,474 @@
...
+                {
+                    for(sai_uint32_t ind = 0; ind < policer_attribs.size(); ind++)
+                    {
+                        auto policer_attr = policer_attribs[ind];
+                        sai_status = sai_policer_api->set_policer_attribute(policer_id, &policer_attr);
+                        if (SAI_STATUS_SUCCESS != sai_status)
+                        {
+                            SWSS_LOG_ERROR("Failed to apply attribute[%d].id=%d to policer for trap group:%s, error:%d\n", ind, policer_attr.id, key.c_str(), sai_status);
+                            return task_process_status::task_failed;
+                        }
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >                     [](start = 25, length = 20)

remove all trailing whitespaces.

File: orchagent/copporch.cpp [hunk 31]
```diff
@@ -0,0 +1,474 @@
...
+                    SWSS_LOG_WARN("Creating policer for existing Trap group:%llx (name:%s).", m_trap_group_map[key], key.c_str());
+                    if (!createPolicer(key, policer_attribs))
+                    {
+                        return task_process_status::task_failed;
+                    }
+                    SWSS_LOG_DEBUG("Created policer:%llx for existing trap group", policer_id);
+                }
+                else 
+                {
+                    for(sai_uint32_t ind = 0; ind < policer_attribs.size(); ind++)
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >or(sai_uint32_t ind = 0; ind < policer_attribs.size(); ind++) [](start = 21, length = 61)

add a TODO comment here.

/* TODO: We should really only set changed attributes. The changes need to detected either by orch agent submodule or by the orch agent framework. */

File: orchagent/copporch.cpp [hunk 32]
```diff
@@ -0,0 +1,474 @@
...
+    SWSS_LOG_DEBUG("trap group id:%llx", m_trap_group_map[key]);
+    if (m_trap_group_policer_map.find(m_trap_group_map[key]) == m_trap_group_policer_map.end())
+    {
+        return SAI_NULL_OBJECT_ID;
+    }
+    SWSS_LOG_DEBUG("trap group policer id:%llx", m_trap_group_policer_map[m_trap_group_map[key]]);
+    return m_trap_group_policer_map[m_trap_group_map[key]];
+}
+
+bool CoppOrch::createPolicer(string trap_group, std::vector<sai_attribute_t> &policer_attribs)
```
Comment (lguohan) [inline orchagent/copporch.cpp]: rename to trap_group_name

File: orchagent/copporch.cpp [hunk 33]
```diff
@@ -0,0 +1,474 @@
...
+    if (SAI_STATUS_SUCCESS != sai_status)
+    {
+        SWSS_LOG_ERROR("Failed to remove policer:%llx, error:%d\n", policer_id, sai_status);
+        return false;
+    }
+    m_trap_group_policer_map.erase(m_trap_group_map[trap_group_name]);    
+    return true;
+}
+
+sai_object_id_t CoppOrch::getPolicer(string key)
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >key [](start = 44, length = 3)

rename to trap_group_name

File: orchagent/copporch.cpp [hunk 34]
```diff
@@ -0,0 +1,474 @@
...
+    else if (op == DEL_COMMAND)
+    {
+        // delete trap group and its policer.
+        if (!removePolicer(key))
+        {
+            SWSS_LOG_ERROR("Failed to remove policer from trap group:%s\n", key.c_str());
+            return task_process_status::task_failed;
+        }
+        sai_status = sai_hostif_api->remove_hostif_trap_group(m_trap_group_map[key]);
+        if (SAI_STATUS_SUCCESS != sai_status)
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >SAI_STATUS_SUCCESS != sai_status [](start = 12, length = 32)

swap this, compiler now is smart enough to detect '=' error.

**Pull request merged**